PLANNER-2535 Deprecate from(...) in favor of forEach(...) (#439)
kiegroup/optaplanner-website: optaplanner-website-root/content/download/upgradeRecipe/upgradeRecipe8.adoc

=== Rename optaplanner.solver.solve-length to optaplanner.solver.solve.duration

The Micrometer meter `optaplanner.solver.solve-length` has been renamed to `optaplanner.solver.solve.duration` to follow Micrometer naming conventions and make it appear better in a variety of monitoring systems. References to `optaplanner.solver.solve-length.+++*+++` in the monitoring system should be changed to `optaplanner.solver.solve.duration.+++*+++`.

== From 8.13.0.Final to 8.14.0.Final

[.upgrade-recipe-major]
=== Constraint Streams: use `forEach(...)` instead of `from(...)`

We have deprecated the following methods on the `ConstraintFactory` interface:

* `from(...)` in favor of `forEach(...)`,
* `fromUnfiltered(...)` in favor of `forEachIncludingNullVars(...)`,
* `fromUniquePair(...)` in favor of `forEachUniquePair(...)`.

> The new methods behave differently when it comes to nullable planning variables.

Reviewer: For clarity, I'd state it like this, to avoid anyone not explicitly doing that thinking this might apply to them:
```suggestion
The new methods behave differently when it comes to `nullable=true` planning variables.
```

> The deprecated methods will eventually be removed in OptaPlanner 9.

Reviewer: Messaging wise, it's not a good idea to mention "OptaPlanner 9" in a way some users could think that it's already planned for a target date (which it is not). Because when users start thinking it's coming soon (which it is not, work hasn't started on it and won't start for years), it creates fear, uncertainty and doubt for the current major version (this is real, however stupid it sounds) and they tend to go with other solutions than OptaPlanner.

How to avoid that? Use an "a" to indicate we don't know yet when (as in target date) this will be.
```suggestion
The deprecated methods will eventually be removed in a future major version.
```

====  Impact on users without nullable planning variables

> Users who do not use over-constrained planning (where `@PlanningVariable(..., nullable = false)`) may simply replace all the uses of the deprecated methods.

Reviewer: (soft) more concise suggestion that makes "Users who do not" more "active tense":
```suggestion
If you do not use over-constrained planning (where `@PlanningVariable(..., nullable = false)`) simply replace all the uses of the `from*` methods with `forEach*` methods.
```

Before in `*ConstraintProvider.java*`:

[source,java]
----

    ...

> 

Reviewer: (soft) code style
I'd remove the "..." before and after the method (unless the class line is shown too), as other upgrade recipe notes that use a method level snippet without surrounding class don't do it either.

    Constraint roomConflict(ConstraintFactory constraintFactory) {
        return constraintFactory.from(Lesson.class)
            ...
    }

    ...

----

After in `*ConstraintProvider.java*`:

[source,java]
----

    ...

    Constraint roomConflict(ConstraintFactory constraintFactory) {
        return constraintFactory.forEach(Lesson.class)
            ...
    }

    ...

----

====  Impact on users with nullable planning variables

> Users who use over-constrained planning (where `@PlanningVariable(..., nullable = true)`) need to take into account the following:

Reviewer: Same principle:
```suggestion
If you use over-constrained planning (where `@PlanningVariable(..., nullable = true)`) take into account the following:
```

> * `forEach(...)` will filter out planning entities with `null` values in any of their genuine planning variables,

Reviewer: Avoid future tense ("will")
```suggestion
* `forEach(...)` filters out planning entities with `null` values in any of their genuine planning variables,
```

regardless if they are initialized or not.
* The same applies for `join(...)` following a `forEach(...)`.
* And the same also applies for conditional propagation (`ifExists(...)`, `ifNotExists(...)` etc.) following a `forEach(...)`.

In practice, what this means is:

* Where you previously used `null` checks to exclude entities with unassigned variables,

> you may remove these null checks after replacing `from(...)` with `forEach(...)`.

Reviewer: Let's nudge them to the right thing :)
```suggestion
you should remove these null checks after replacing `from(...)` with `forEach(...)`.
```

[source,java]
----

    ...

    Constraint departmentSpecialismConstraint(ConstraintFactory constraintFactory) {
        return constraintFactory.from(BedDesignation.class)
                .filter(bd -> bd.getBed() != null)
                ...
    }

    ...

----

After in `*ConstraintProvider.java*`:

[source,java]
----

    ...

    Constraint departmentSpecialismConstraint(ConstraintFactory constraintFactory) {
        return constraintFactory.forEach(BedDesignation.class)
                ...
    }

    ...

----

> * Where you deliberately intend to retrieve entities with `null` variables,

Reviewer: Let's add the typical case where this happens. This will cover 99% of the migrations:
```suggestion
* Where you deliberately intend to retrieve entities with `null` variables,
for example in the medium constraint to penalize all unassigned entities,
```

> you have to replace `from(...)` with `forEachIncludingNullVars(...)`.

Reviewer: nitpick: "must" in favor of "have to", as it easier to spot when reading diagonally.
```suggestion
you must replace `from(...)` with `forEachIncludingNullVars(...)`.
```

[source,java]
----

    ...

    Constraint femaleInMaleRoomConstraint(ConstraintFactory constraintFactory) {
        return constraintFactory.from(BedDesignation.class)
                .filter(bd -> bd.getPatientGender() == Gender.FEMALE
                        && bd.getRoomGenderLimitation() == GenderLimitation.MALE_ONLY)

>                 ...

Reviewer: (hard)
This isn't an actual constraint that needs to be migrated to forEachIncludingNullVars(). It's fake. Very confusion.

Let's use this one instead. Including the penalize method:

```
    public Constraint assignEveryPatientToABedConstraint(ConstraintFactory constraintFactory) {
        return constraintFactory.forEachIncludingNullVars(BedDesignation.class)
                .filter(bd -> bd.getBed() == null)
                .penalize("assignEveryPatientToABed", HardMediumSoftScore.ONE_MEDIUM,
                        BedDesignation::getAdmissionPartNightCount);
    }
```
